mars/oscar/backends/ray/communication.py

```diff
@@ -0,0 +1,282 @@
+
+    @implements(Channel.recv)
+    async def recv(self):
+        if self._closed.is_set():  # pragma: no cover
+            raise ChannelClosed('Channel already closed, cannot write message')
+        try:
+            # Wait on ray object ref
+            object_ref = await self._in_queue.get()
+            return await object_ref
+        except RuntimeError as e:
+            if not self._closed.is_set():
+                raise e
+
+
+class RayServerChannel(RayChannelBase):
+    """A channel from ray actor to ray driver/actor. Since ray actor can't call ray driver,
+     we use ray call reply for server channel send. Note that there can't be multiple
+     channel message sends for one received message, or else it will be taken as next
+     message's reply.
+    """
+    __slots__ = '_out_queue', '_msg_recv_counter', '_msg_sent_counter'
+
+    def __init__(self,
+                 local_address: str = None,
+                 channel_index: int = None,
+                 channel_id: ChannelID = None,
+                 compression=None):
+        super().__init__(local_address, None, channel_index, channel_id, compression)
+        self._out_queue = asyncio.Queue()
+        self._msg_recv_counter = 0
+        self._msg_sent_counter = 0
+
+    @implements(Channel.send)
+    async def send(self, message: Any):
+        if self._closed.is_set():  # pragma: no cover
+            raise ChannelClosed('Channel already closed, cannot send message')
+        # Current process is ray actor, we use ray call reply to send message to ray driver/actor.
+        # Not that we can only send once for every read message in channel, otherwise
+        # it will be taken as other message's reply.
+        await self._out_queue.put(message)
+        self._msg_sent_counter += 1
+        assert self._msg_sent_counter <= self._msg_recv_counter, \
+            "RayServerChannel channel doesn't support send multiple replies for one message."
+
+    @implements(Channel.recv)
+    async def recv(self):
+        if self._closed.is_set():  # pragma: no cover
+            raise ChannelClosed('Channel already closed, cannot write message')
+        try:
+            return await self._in_queue.get()
+        except RuntimeError:
+            if self._closed.is_set():
+                pass
+
+    async def __on_ray_recv__(self, message):
+        """This method will be invoked when current process is a ray actor rather than a ray driver"""
+        self._msg_recv_counter += 1
+        await self._in_queue.put(message)
+        task = asyncio.create_task(self._out_queue.get())
+        while True:  # pragma: no cover
+            # Avoid hang when channel is closed after `self._out_queue.get()` is awaited.
+            done, pending = await asyncio.wait([task], timeout=0.1)
+            if self._closed.is_set():
+                raise ChannelClosed('Channel already closed')
+            if done:
+                return await task
```

Comment:
Is it simpler to ?
```python
done, pending = asyncio.wait([self._out_queue.get(), self._closed.wait()], return_when=asyncio.FIRST_COMPLETED)
if self._closed.is_set():
    raise ChannelClosed('Channel already closed')
else:
    return await done[0]  # May be it's safer to check that the done is the `self._out_queue.get()`.
```

Reply:
`self._out_queue.get()` will never finish when When using `asyncio.FIRST_COMPLETED` and there is an exception throwed in `channel send` or `channel_handler` , `self._out_queue.get()` will never return. That's why we use `wait` and `timeout`.
`asyncio.FIRST_COMPLETED` here has same effect with `await self._out_queue.get()` which is we try to avoid here.

Reply:
How do you check `May be it's safer to check that the done is the `self._out_queue.get()`.`?

Reply:
``` python
# pseudo-code
task = create_task(self._out_queue.get())
done, pending = asyncio.wait([task, self._closed.wait()], ...)
# check self._close.is_set
if task in done:
    return await task.
else:
    raise Excepion
```

But, just `await done[0]` is OK. if both of `self._closed.wait()` and `self._out_queue.get()` are done, then we raises ChannelClosed('Channel already closed').